File: day05/day05.go
package day05

import (
	"fmt"
	"strconv"
	"strings"
)

func parseStacks(input string) [][]byte {
	stacks := make([][]byte, 0)
	stackLines := make([]string, 0)
	stackCount := 0

	for _, l := range strings.Split(input, "\n") {
		if strings.HasPrefix(l, "[") {
			stackLines = append(stackLines, l)
		} else {
			stackCount = len(strings.Replace(l, " ", "", -1))
			break
		}
	}
	fmt.Printf("creating %d stacks from following lines:\n", stackCount)
	for _, line := range stackLines {
		fmt.Println(line)
	}
	for i := 0; i < stackCount; i++ {
		stacks = append(stacks, []byte{})
	}
	for i := len(stackLines) - 1; i >= 0; i-- {
		line := stackLines[i]
		for j := 0; j < stackCount; j++ {
			item := line[1+j*4]
			if item != ' ' {
				stacks[j] = append(stacks[j], item)
			}
		}
	}
	return stacks
}

func parseInput(input string) ([][]byte, [][3]int) {
	stacks := parseStacks(input)
	instructions := parseInstructions(input)
	return stacks, instructions
}

func parseInstructions(input string) [][3]int {
	instructions := make([][3]int, 0)
	for _, l := range strings.Split(input, "\n") {
		if strings.HasPrefix(l, "move") {
			parts := strings.Split(l, " ")
			inst := [3]int{}
			inst[0], _ = strconv.Atoi(parts[1])
			inst[1], _ = strconv.Atoi(parts[3])
			inst[1]--
			inst[2], _ = strconv.Atoi(parts[5])
			inst[2]--
			instructions = append(instructions, inst)
		}
	}
	return instructions
}

func doMoves(stacks [][]byte, instructions [][3]int, moveAtOnce bool) {
	for _, inst := range instructions {
		count := inst[0]
		fromStack := inst[1]
		toStack := inst[2]
		if moveAtOnce {
			items := stacks[fromStack][len(stacks[fromStack])-count:]
			stacks[fromStack] = stacks[fromStack][:len(stacks[fromStack])-count]
			stacks[toStack] = append(stacks[toStack], items...)
		} else {
			for i := 0; i < count; i++ {
				item := stacks[fromStack][len(stacks[fromStack])-1]
				stacks[fromStack] = stacks[fromStack][:len(stacks[fromStack])-1]
				stacks[toStack] = append(stacks[toStack], item)
			}
		}
	}
}

func Solve(input string, test bool, task int) (string, string) {
	res1, res2 := "", ""
	stacks, instructions := parseInput(input)
	stacks2 := make([][]byte, len(stacks))
	for i := range stacks {
		stacks2[i] = make([]byte, len(stacks[i]))
		copy(stacks2[i], stacks[i])
	}

	// move one item at a time
	if task != 2 {
		doMoves(stacks, instructions, false)
		res1 = ""
		for _, s := range stacks {
			if len(s) > 0 {
				res1 += string(s[len(s)-1])
			}
		}
		fmt.Println("Task 1: ", res1)
	}

	if task != 1 {
		// move all items at once
		doMoves(stacks2, instructions, true)
		res2 = ""
		for _, s := range stacks2 {
			if len(s) > 0 {
				res2 += string(s[len(s)-1])
			}
		}
		fmt.Println("Task 2: ", res2)
	}

	return res1, res2
}

File: aoc22_rust/src/day05.rs
use crate::types::Task;

fn parse_stacks(input: &str, test: bool) -> Vec<Vec<u8>> {
    let mut stack_lines = vec![];
    let mut stack_count = 0;

    for l in input.lines() {
        if l.trim_start().starts_with("[") {
            stack_lines.push(l);
        } else {
            stack_count = l.bytes().filter(|b| *b != b' ').count();
            break;
        }
    }
    if test {
        println!("creating {stack_count} stacks from following lines:");
        for line in &stack_lines {
            println!("{line}");
        }
    }
    let mut stacks = vec![];
    for _ in 0..stack_count {
        stacks.push(vec![]);
    }
    for line in stack_lines.iter().rev() {
        for j in 0..stack_count {
            let item = line.as_bytes()[1 + j * 4];
            if item != b' ' {
                stacks[j].push(item)
            }
        }
    }
    return stacks;
}

fn parse_input(input: &str, test: bool) -> (Vec<Vec<u8>>, Vec<[usize; 3]>) {
    let stacks = parse_stacks(input, test);
    if test {
        println!("parsed following stacks:");
        for s in &stacks {
            println!("{s:?}");
        }
    }
    let instructions = parse_instructions(input);
    if test {
        println!("parsed following instructions:\n{instructions:?}");
    }
    return (stacks, instructions);
}

fn parse_instructions(input: &str) -> Vec<[usize; 3]> {
    let mut instructions = vec![];
    for l in input.lines() {
        if l.starts_with("move") {
            let parts: Vec<&str> = l.split(" ").collect();
            let mut inst = [0; 3];
            inst[0] = parts[1].parse::<usize>().unwrap();
            inst[1] = parts[3].parse::<usize>().unwrap();
            inst[1] -= 1;
            inst[2] = parts[5].parse::<usize>().unwrap();
            inst[2] -= 1;
            instructions.push(inst);
        }
    }
    return instructions;
}

fn do_moves(
    stacks: &mut Vec<Vec<u8>>,
    instructions: &Vec<[usize; 3]>,
    move_at_once: bool,
    _test: bool,
) {
    for inst in instructions {
        let count = inst[0];
        let from_stack = inst[1];
        let to_stack = inst[2];
        if move_at_once {
            let mut items = stacks[from_stack][stacks[from_stack].len() - count..].to_owned();
            let stack_len = stacks[from_stack].len();
            stacks[from_stack].truncate(stack_len - count);
            stacks[to_stack].append(&mut items);
        } else {
            for _i in 0..count {
                let item = stacks[from_stack].pop().unwrap();
                stacks[to_stack].push(item);
            }
        }
    }
}

pub fn solve(input: &str, test: bool, task: Task) -> (String, String) {
    let mut res1 = String::from("");
    let mut res2 = String::from("");
    let (mut stacks, instructions) = parse_input(input, test);
    let mut stacks2 = stacks.clone();

    // move one item at a time
    if !matches!(task, Task::Two) {
        do_moves(&mut stacks, &instructions, false, test);
        res1 = "".into();
        for s in stacks {
            if s.len() > 0 {
                res1.push_str(&format!("{}", *s.last().unwrap() as char));
            }
        }
    }

    if !matches!(task, Task::One) {
        // move all items at once
        do_moves(&mut stacks2, &instructions, true, test);
        res2 = "".into();
        for s in stacks2 {
            if s.len() > 0 {
                res2.push_str(&(*s.last().unwrap() as char).to_string());
            }
        }
    }

    return (res1, res2);
}
